### ISSUE TITLE
Provide host based Ansible provisioner

### ISSUE THREAD

**Issue by bill-myers** [mitchellh/packer#1555]
The current Ansible provisioner runs Ansible inside the created box, which is totally absurd.

Among other things, it means you can't copy local files to the box using Ansible, which is pretty much the main purpose of provisioning.

Please provide a proper provisioner that runs ansible on the host machine and tells it to connect to the created box via ssh.

**Comment by peterloron** [comment]
+1

@bhcleek Any progress update? Thanks.

**Comment by bhcleek** [comment]
Thanks for asking about this @peterloron. I haven't been able to make meaningful progress yet (the end of the year got a bit hectic), but I am getting back to it this week. Your query prompted me to do some analysis that helped me resolve some of the blockers I had.

**Comment by tbaylis** [comment]
:+1: This would be really useful!

**Comment by peterloron** [comment]
Cool! Let me know if I can help out. I'm not a Go whiz, but I may be 
able to help with the code, testing, docs, etc.

-Pete

**Comment by bhcleek** [comment]
I spent some time on this last night trying out different solutions. I may have to provide a provisioner that's not ready to be integrated into Packer yet, because the only solutions that will work are not compatible with all the builders.

Short of modifying the Communicator interface to have a RemoteAddr() method that provides the IP address of the client, which could then at least be checked for nil before trying to launch ansible, the most easily implemented solution that I can see is to type assert the communicator and then reflect to to get the remote address. That's definitely not The Go Way&#x2122; and is definitely not the way Packer wants provisioners to work.

There's another way through that I see, too, but it may be tricky to pull off. First, create a new SSH server on the host using golang.org/x/crypto/ssh, then proxy data from that request to the communicator passed to the provisioner. This would require that the SSH proxy understand and deal with the types of requests that Ansible may send. From looking at the ansible code, I think it's all either EXEC or SCP requests. This is the cleaner solution, but also more error prone. I thinIt has the advantage, though, of having the best chance of getting merged

I'm open to other ideas if anyone wants to contribute them.

Here's a bit of background reading, too: https://groups.google.com/forum/#!searchin/packer-tool/ansible/packer-tool/Go_nAVgkt2g/nWW1SvCruiAJ

**Comment by peterloron** [comment]
A quick glance at the Packer code makes me think adding that 
RemoteAddr() method would be easy and minimally disruptive. Yes, it's 
touching a core bit of code, but the change is additive and not 
modifying any existing method. The ansible-remote (or whatever) 
provisioner would similarly be encapsulated and additive. It's also much 
more elegant than trying to do a reflector or proxy lashup.

Perhaps it's worth pinging Mitchell to see if he would consider a pull 
request against Communicator?

-Pete

**Comment by bhcleek** [comment]
Not all builders use SSH, though, so if RemoteAddr() == nil, the provisioner would have to bail out, and that doesn't seem acceptable. As a user of Packer, I wouldn't expect that not all combinations of builders and provisioners are supported.

**Comment by peterloron** [comment]
Hmmm. Perhaps a general "ansible" provisioner which would fail to the existing "ansible-local" when no RemoteAddr is present? This would avoid breaking existing installs, and is somewhat analogous to how the vmware-* provisioners work in that you can direct them to connect to an ESXi server or to run locally.

**Comment by bhcleek** [comment]
I may give that a shot if proxying an SSH connection from Ansible the Communicator doesn't work out, but I'm going to have a go at trying to do this without requiring changes to the Communicator interface and without using reflection.

**Comment by bhcleek** [comment]
Just to give an update on this: I've made good progress this week and am in the middle of testing and refactoring.

**Comment by peterloron** [comment]
Fantastic! I look forward to taking it for a spin.

**Comment by bhcleek** [comment]
I've created a provisioner that supports establishing a connection using paramiko. The provisioner still has a good amount of work to be done: supporting native OpenSSH, and both SCP and SFTP support for file transfers.

The provisioner is still very young, and is not yet useful, but you can follow along with the progress I'm making if you want: https://github.com/bhcleek/packer-provisioner-ansible

**Comment by rbraley** [comment]
Thanks Billie,




This is good news! Is there anything I can do to help testing it? 


Ryan Braley  |  Founder 
http://traintracks.io/
US: +1 (206) 866 5661



CN: +86 185 1129 5661



Coding the future. Decoding the game.

**Comment by bhcleek** [comment]
Thanks, @rbraley, but I think it's still too early to start testing; the provisioner is so young that there wouldn't be much point yet.

**Comment by rbraley** [comment]
Ok, do let me know if I can help in some way :D


Ryan Braley  |  Founder 
http://traintracks.io/
US: +1 (206) 866 5661



CN: +86 185 1129 5661



Coding the future. Decoding the game.

**Comment by bhcleek** [comment]
[bhcleek/packer-provisioner-ansible](http://github.com/bhcleek/packer-provisioner-ansible) works without requiring Ansible to be installed on the machine being provisioned. I'm interested in getting some feedback to help closing any functional gaps before I submit a PR here. **it is dependent on #1920**

**Comment by bhcleek** [comment]
I realized that bhcleek/packer-provisioner-ansible, strictly speaking, is not actually dependent on #1920. Rather, it is simply dependent on being able to execute an sftp-server on the machine being provisioned without writing anything to stdout when the sftp-server is started.

Since the SSH communicator currently allocates a PTY when `ssh.Start(*RemoteCmd)` is called, most environments will violate that expectation.

**Comment by poolpog** [comment]
I've used two workarounds to good effect.

1) Shell provisioner to clone the Ansible repo. This took some finagling because of the permissions on the repo, ssh key setup, etc., but worked.

2) Create a tarball of the repo locally and upload, then untar it, using the file provisioner and the shell provisioner.

**Comment by dkoppstein** [comment]
Before going down this rabbit hole, there's a good design reason for why it's done this way. For long, complex installations, the ssh connection can time out and you're left with an incomplete box. You can work around this with asynchronous tasks to some extent for simple playbooks, but it doesn't work for a lot of things (see ansible/ansible#5841). The most robust way, imo, is to do it the Packer way, which is to just execute it as a local ansible playbook, walk away, and let it do its thing.  

If you don't want ansible in the image, until ansible gets asynchronous execution of entire roles or playbooks, I think it's best to just uninstall it as a cleanup task. 

It would, however, be nice to have Packer install/uninstall ansible instead of having to do it yourself, and with a shiny new ansible, not the 1.5.x junk that comes with apt by default.

**Comment by pas256** [comment]
Well said @dkoppstein

**Comment by dkoppstein** [comment]
Actually, I want to amend my statement. It might be possible to robustly provision with ansible with a few tweaks.  

Maybe something like 

1. Disable requiretty in /etc/sudoers if enabled
2. provision ansible remotely with pipelining=True, ssh_args = -o ControlPersist=1800s
3. Re-enable requiretty if necessary

I haven't actually written any code, just thinking out loud =)

**Comment by bhcleek** [comment]
@dkoppstein That may work when packer is communicating with the machine via ssh, but it wouldn't work with the Docker provisioner, for instance.

**Comment by dkoppstein** [comment]
Unfortunately not :(

It does, however, work very well with [this](https://github.com/ansible/immutablish-deploys/blob/master/build_ami.yml).

**Comment by bhcleek** [comment]
Yeah, as long as the packer is communicating with the machine, the approach you suggested would work. I wanted something that would allow me to provision with Ansible without regard for the provider being used, so I created #1969 .

**Comment by pikeas** [comment]
Why has @bhcleek's PR been left hanging without comment?
